src/util/RtreeBuild.cpp

```diff
@@ -0,0 +1,1133 @@
+    for (RTreeValue element : sorterRectsD0Basic.sortedView()) {
+      RTreeValueWithOrderIndex entry = {element.box, element.id, xSize, 0};
+      sorterRectsD1.push(entry);
+      xSize++;
+
+      if (globalMinX == -1 || element.box.min_corner().get<0>() < globalMinX) {
+        globalMinX = element.box.min_corner().get<0>();
+      }
+      if (globalMinY == -1 || element.box.min_corner().get<1>() < globalMinY) {
+        globalMinY = element.box.min_corner().get<1>();
+      }
+      if (element.box.max_corner().get<0>() > globalMaxX) {
+        globalMaxX = element.box.max_corner().get<0>();
+      }
+      if (element.box.max_corner().get<1>() > globalMaxY) {
+        globalMaxY = element.box.max_corner().get<1>();
+      }
+    }
+  }
+  sorterRectsD0Basic.clear();
+
+  size_t currentS = std::ceil(((float) xSize) / ((float) M));
+
+  uint64_t ySize = 0;
+  std::ofstream r1File = std::ofstream(onDiskBase + ".boundingbox.d1.tmp", std::ios::binary);
+  ad_utility::BackgroundStxxlSorter<RTreeValueWithOrderIndex, SortRuleLambdaXWithIndex> sorterRectsD0 = ad_utility::BackgroundStxxlSorter<RTreeValueWithOrderIndex, SortRuleLambdaXWithIndex>(std::ceil((maxBuildingRamUsage < 9999999999 ? maxBuildingRamUsage : 9999999999) / 3.0));
+  std::shared_ptr<multiBoxWithOrderIndex> RectanglesD0WithOrder = std::make_shared<multiBoxWithOrderIndex>();
+  std::shared_ptr<multiBoxWithOrderIndex> r1Small = std::make_shared<multiBoxWithOrderIndex>();
+  // placeholder
+  r1Small->push_back(RTreeValueWithOrderIndex());
+  r1Small->push_back(RTreeValueWithOrderIndex());
+  RTreeValueWithOrderIndex minD1;
+  RTreeValueWithOrderIndex maxD1;
+
+  if (workInRam) {
+    for (RTreeValueWithOrderIndex element : *RectanglesD1WithOrder) {
+      element.orderY = ySize;
+      RectanglesD0WithOrder->push_back(element);
+
+      if (((ySize + 1) % currentS == 0 && (ySize + 1) / currentS >= 1 && (ySize + 1) / currentS < M)
+          || (ySize % currentS == 0 && ySize / currentS >= 1 && ySize / currentS < M)) {
+        // index i * S - 1 or i * S
+        r1Small->push_back(element);
+      }
+
+      if (ySize == 0) {
+        minD1 = element;
+        maxD1 = element;
+      }
+      if (element.orderY > maxD1.orderY) {
+        maxD1 = element;
+      }
+
+      ySize++;
+    }
+    centerOrdering(*RectanglesD0WithOrder, 0);
+  } else {
+    for (RTreeValueWithOrderIndex element : sorterRectsD1.sortedView()) {
+      element.orderY = ySize;
+      Rtree::SaveEntryWithOrderIndex(element, r1File);
+      sorterRectsD0.push(element);
+
+      if (((ySize + 1) % currentS == 0 && (ySize + 1) / currentS >= 1 && (ySize + 1) / currentS < M)
+          || (ySize % currentS == 0 && ySize / currentS >= 1 && ySize / currentS < M)) {
+        // index i * S - 1 or i * S
+        r1Small->push_back(element);
+      }
+
+      if (ySize == 0) {
+        minD1 = element;
+        maxD1 = element;
+      }
+      if (element.orderY > maxD1.orderY) {
+        maxD1 = element;
+      }
```

Comment:
Is it correct that the `min` and `max` should also be the `first` and the `last` value?
I have a nice idea on how to implement this whole loop using C++20 ranges and views (have a look at them):
You first to a `views::transform` that pushes whatever needs to be pushed and returns the argument unchanged. On the resulting range you call `ranges::min::max`.